pyomo/gdp/plugins/bigm.py

```diff
@@ -253,117 +224,125 @@ def _apply_to_impl(self, instance, **kwds):
 
-    def _transformDisjunction(self, obj, transBlock, bigM):
+    def _transform_disjunction(self, obj, bigM):
+        transBlock = self._add_transformation_block(obj.parent_block())
+
+        # If this is an IndexedDisjunction, we have to create the XOR constraint
+        # here because we want its index to match the disjunction. In any case,
+        # we might as well.
+        xorConstraint = self._get_xor_constraint(obj)
+
         # relax each of the disjunctionDatas
         for i in sorted(iterkeys(obj)):
-            self._transformDisjunctionData(obj[i], transBlock, bigM, i)
+            self._transform_disjunctionData(obj[i], bigM, i, xorConstraint,
+                                            transBlock)
 
-        # deactivate so we know we relaxed
+        # deactivate so the writers don't scream
         obj.deactivate()
 
-    def _transformDisjunctionData(self, obj, transBlock, bigM, index):
+    def _transform_disjunctionData(self, obj, bigM, index, xorConstraint=None,
+                                   transBlock=None):
         if not obj.active:
-            return  # Do not process a deactivated disjunction
-        parent_component = obj.parent_component()
-        transBlock.disjContainers.add(parent_component)
-        orConstraint = self._getXorConstraint(parent_component)
+            return  # Do not process a deactivated disjunction 
+        # We won't have these arguments if this got called straight from
+        # targets. But else, we created them earlier, and have just been passing
+        # them through.
+        if transBlock is None:
+            transBlock = self._add_transformation_block(obj.parent_block())
+        if xorConstraint is None:
+            xorConstraint = self._get_xor_constraint(obj.parent_component())
 
         xor = obj.xor
         or_expr = 0
+        # Just because it's unlikely this is what someone meant to do...    
+        if len(obj.disjuncts) == 0:
+            raise GDP_Error("Disjunction %s is empty. This is " 
+                            "likely indicative of a modeling error."  %
+                            obj.getname(fully_qualified=True,
+                                        name_buffer=NAME_BUFFER))
         for disjunct in obj.disjuncts:
             or_expr += disjunct.indicator_var
             # make suffix list. (We don't need it until we are
             # transforming constraints, but it gets created at the
             # disjunct level, so more efficient to make it here and
-            # pass it down.
+            # pass it down.)
             suffix_list = self._get_bigm_suffix_list(disjunct)
             # relax the disjunct
-            self._bigM_relax_disjunct(disjunct, transBlock, bigM, suffix_list)
+            self._transform_disjunct(disjunct, transBlock, bigM, suffix_list)
+
         # add or (or xor) constraint
         if xor:
-            orConstraint.add(index, (or_expr, 1))
+            xorConstraint.add(index, (or_expr, 1))
```

Comment:
Can we do `expr=or_expr == 1`, or is the tuple for efficiency purposes?

Reply:
I don't know?? @jsiirola?

Reply:
The tuple notation is slightly faster, but not enough to warrant using it at the cost of clarity.